File: study1/week2/invert-binary-tree.java
// https://leetcode.com/problems/invert-binary-tree/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    // Time Complexity: O(n)
    // Space Complexity: O(1)
    public TreeNode invertTree(TreeNode root) {
        swapNode(root);
        traverseTree(root);

        return root;
    }

    public void traverseTree(TreeNode node) {
        if (node == null) {
            return;
        }
        swapNode(node.left);
        traverseTree(node.left);
        swapNode(node.right);
        traverseTree(node.right);
    }

    public void swapNode(TreeNode node) {
        if (node == null) {
            return;
        }

        TreeNode temp = node.left;
        node.left = node.right;
        node.right = temp;
    }
}

File: study1/solutions/week2/invert_binary_tree.py
import queue

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return root
        toVisit = queue.Queue()
        toVisit.put(root)
        while not toVisit.empty():
            current = toVisit.get()
            current.left, current.right = current.right, current.left
            if current.left is not None:
                toVisit.put(current.left)
            if current.right is not None:
                toVisit.put(current.right)
        return root